s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/utils/SourceRecordIterator.java

```diff
@@ -16,215 +16,165 @@
-
-    private final OffsetManager offsetManager;
-
+    /** The OffsetManager that we are using */
+    private final OffsetManager<S3OffsetManagerEntry> offsetManager;
+    /** The offset manager Entry we are working with */
+    private S3OffsetManagerEntry offsetManagerEntry;
+    /** The Configuration that we are using */
     private final S3SourceConfig s3SourceConfig;
-    private final String bucketName;
-
+    /** The S3Object iterator that we read from */
+    private final Iterator<S3Object> s3ObjectIterator;
+    /** The transformer we apply to the data */
     private final Transformer transformer;
-    // Once we decouple the S3Object from the Source Iterator we can change this to be the SourceApiClient
-    // At which point it will work for al our integrations.
+    /** The Source client we are reading from */
     private final AWSV2SourceClient sourceClient; // NOPMD
-
-    public SourceRecordIterator(final S3SourceConfig s3SourceConfig, final OffsetManager offsetManager,
-            final Transformer transformer, final AWSV2SourceClient sourceClient) {
+    /** THe iterator we will return data from */
+    private Iterator<S3SourceRecord> outerIterator;
+
+    /**
+     * Constructor.
+     *
+     * @param s3SourceConfig
+     *            The configuration.
+     * @param offsetManager
+     *            the offset manager.
+     * @param transformer
+     *            the transformer to sue.
+     * @param sourceClient
+     *            the source client to read from.
+     */
+    public SourceRecordIterator(final S3SourceConfig s3SourceConfig,
+            final OffsetManager<S3OffsetManagerEntry> offsetManager, final Transformer transformer,
+            final AWSV2SourceClient sourceClient) {
         this.s3SourceConfig = s3SourceConfig;
         this.offsetManager = offsetManager;
-
-        this.bucketName = s3SourceConfig.getAwsS3BucketName();
         this.transformer = transformer;
         this.sourceClient = sourceClient;
-        objectListIterator = sourceClient.getListOfObjectKeys(null);
-    }
-
-    private void nextS3Object() {
-        if (!objectListIterator.hasNext()) {
-            // Start after the object Key we have just finished with.
-            objectListIterator = sourceClient.getListOfObjectKeys(currentObjectKey);
-            if (!objectListIterator.hasNext()) {
-                recordIterator = Collections.emptyIterator();
-                return;
-            }
-        }
-
-        try {
-            currentObjectKey = objectListIterator.next();
-            if (currentObjectKey != null) {
-                recordIterator = createIteratorForCurrentFile();
-            }
-        } catch (IOException e) {
-            throw new AmazonClientException(e);
-        }
+        this.s3ObjectIterator = IteratorUtils.filteredIterator(sourceClient.getIteratorOfObjects(null),
+                s3Object -> extractOffsetManagerEntry(s3Object));
```

Comment:
Lambda can be replaced with method reference 
```suggestion
                this::extractOffsetManagerEntry);
```